# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLContext.h
# Commit: e030be1d2dd9
# Full Hash: e030be1d2dd9b0bb47680fac29f20447c9d66f8c
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-02-21 09:51:10
# Regressor Bug: 1616715
# File Overlap Count: 1
# Description:
#   Bug 1616715 - Make ReadPixelDesc per-call. r=lsalzman
#   
#   GetSurfaceSnapshot (and similar) want to ReadPixels the backbuffer,
#   without having to navigate the GL PIXEL_PACK state.
#   Since this is cold cold code, it's nice to reduce the backend state by
# ==============================================================================

diff -r adc841b3e204 -r e030be1d2dd9 dom/canvas/WebGLContext.h
--- a/dom/canvas/WebGLContext.h	Fri Feb 21 02:27:08 2020 +0000
+++ b/dom/canvas/WebGLContext.h	Fri Feb 21 03:12:41 2020 +0000
@@ -610,7 +610,7 @@
 
   void LineWidth(GLfloat width);
   void LinkProgram(WebGLProgram& prog);
-  void PixelStorei(GLenum pname, GLint param);
+  void PixelStorei(GLenum pname, uint32_t param);
   void PolygonOffset(GLfloat factor, GLfloat units);
 
   RefPtr<layers::SharedSurfaceTextureClient> GetVRFrame();
@@ -623,20 +623,15 @@
       const webgl::FormatUsageInfo* usage) const;
 
  protected:
-  void ReadPixelsImpl(GLint x, GLint y, GLsizei width, GLsizei height,
-                      GLenum format, GLenum type, uintptr_t data,
+  void ReadPixelsImpl(const webgl::ReadPixelsDesc&, uintptr_t data,
                       uint64_t dataLen);
-  bool DoReadPixelsAndConvert(const webgl::FormatInfo* srcFormat, GLint x,
-                              GLint y, GLsizei width, GLsizei height,
-                              GLenum format, GLenum destType, uintptr_t dest,
+  bool DoReadPixelsAndConvert(const webgl::FormatInfo* srcFormat,
+                              const webgl::ReadPixelsDesc&, uintptr_t dest,
                               uint64_t dataLen, uint32_t rowStride);
 
  public:
-  void ReadPixelsPbo(GLint x, GLint y, GLsizei width, GLsizei height,
-                     GLenum format, GLenum type, uint64_t offset);
-
-  void ReadPixels(GLint x, GLint y, GLsizei width, GLsizei height,
-                  GLenum format, GLenum type, const Range<uint8_t>& dest);
+  void ReadPixelsPbo(const webgl::ReadPixelsDesc&, uint64_t offset);
+  void ReadPixels(const webgl::ReadPixelsDesc&, const Range<uint8_t>& dest);
 
   ////
 
@@ -1161,10 +1156,6 @@
   CheckedUint32 GetUnpackSize(bool isFunc3D, uint32_t width, uint32_t height,
                               uint32_t depth, uint8_t bytesPerPixel);
 
-  bool ValidatePackSize(uint32_t width, uint32_t height, uint8_t bytesPerPixel,
-                        uint32_t* const out_rowStride,
-                        uint32_t* const out_endOffset);
-
   UniquePtr<webgl::TexUnpackBlob> FromDomElem(
       const dom::HTMLCanvasElement& canvas, TexImageTarget target, uvec3 size,
       const dom::Element& elem, ErrorResult* const out_error) const;